quickwit-indexing/src/actors/pipeline_supervisor.rs

```diff
@@ -416,7 +414,40 @@ impl IndexingPipelineSupervisor {
             // }
         } else {
             match self.healthcheck() {
-                Health::Healthy => {}
+                Health::Healthy => {
+                    if let Some(handlers) = self.handlers.as_mut() {
+                        // We want to terminate the pipeline once the source is exhausted,
+                        // the indexer has exited and all merge operations have been done.
+                        // This is true when all merging pipeline actors are all in `IDLE` state
+                        // and indexing actors have exited.
+                        if handlers.source.state() == ActorState::Exit
+                            && handlers.indexer.state() == ActorState::Exit
+                            && handlers.packager.state() == ActorState::Exit
+                            && handlers.uploader.state() == ActorState::Exit
+                            && handlers.publisher.state() == ActorState::Exit
+                            && handlers.merge_planner.state() == ActorState::Idle
+                            && handlers.merge_split_downloader.state() == ActorState::Idle
+                            && handlers.merge_executor.state() == ActorState::Idle
```

Comment:
I think there is a small race condition. (Maybe we don't care).
There are many ways for this condition to be true at one point of time, and not true later.

Reply:
Again better than nothing, but I think it is important to be aware of that.

Reply:
Indeed, I was not very confident when I wrote this code, many equality checks on shifting sands.
I will leave it a few days and see if I come up with better ideas or at least some docs to explain the limitations of this approach.

Reply:
To make the shutdown more reliable, here is what we can do when there is only the merge pipeline running:
- keep a boolean to track the fact that the merge pipeline was idle
- on supervise
  - if merge pipeline is idle then: if bool is true, exit if not set bool to true
  - if not, set bool to false

Reply:
@fulmicoton I added a state variable on the indexing pipeline to terminate the pipeline after 2 consecutive idle merge pipeline states.